### PR TITLE
Added optional/default parameters for StreamWriter/StreamReader

### LINKED ISSUE #8173: Simpler way to specify leaveOpen in StreamWriter constructor
Currently, the `StreamWriter` constructor has the following overloads:

``` c#
public StreamWriter(Stream stream)
public StreamWriter(Stream stream, Encoding encoding)
public StreamWriter(Stream stream, Encoding encoding, int bufferSize)
public StreamWriter(Stream stream, Encoding encoding, int bufferSize, bool leaveOpen)
```

I believe a relatively common need is to set `leaveOpen` to `true`, while keeping both `encoding` and `bufferSize` at their default values.

But the process of doing that goes something like this:
- Figure out there is no way to specify just `leaveOpen`.
- From the documentation of `StreamWriter(Stream stream)`, figure out that the default encoding is UTF-8. (Though it doesn't mention it's actually UTF-8 _without BOM_.)
- Either make a guess about what a good buffer size is, or look at the source code of `StreamWriter`.
- Finally write `new StreamWriter(stream, Encoding.UTF8, 1024, true)`.

I don't think that's acceptable, so I'm proposing to make `new StreamWriter(stream, leaveOpen: true)` work.

I can see few options on how to do this:
1. Assuming optional arguments are okay (they probably aren't, see dotnet/corefx#470), make use of them by changing the full overload to:
   
   ``` c#
   public StreamWriter(
       Stream stream, Encoding encoding = null,
       int bufferSize = -1, bool leaveOpen = false)
   ```
   
   This would require relaxing `encoding` and `bufferSize` to accept `null` and `-1`, respectively, but I think that's okay.
   
   I think this is the most convenient option, since it allows one to easily specify any combination of options.
2. Add a new constructor overload:
   
   ``` c#
   public StreamWriter(Stream stream, bool leaveOpen)
   ```
   
   Adding some other overloads might make sense too.
3. Don't add anything, just relax `encoding` and `bufferSize` as in option 1.
   
   I think this is the most conservative option, but also least convenient, since it would require writing `new StreamWriter(stream, null, -1, true)`.

### PR DESCRIPTION
**Pull request by jimdemis**
This resolves #8173
The CoreCLR side: [dotnet/coreclr#24056](https://github.com/dotnet/coreclr/pull/24056)

### CODE REVIEW

File: src/System.IO/tests/StreamReader/StreamReader.CtorTests.cs [hunk 1]
```diff
@@ -47,5 +48,29 @@ public static void CreationFromMemoryStreamWithEncodingTrue()
...
+            using (var sr = new StreamReader(new MemoryStream(), bufferSize: -1))
+            {
+                Assert.Equal(Encoding.UTF8, sr.CurrentEncoding);
+            }
+
+            using (var sr = new StreamReader(new MemoryStream(), detectEncodingFromByteOrderMarks: true))
+            {
+                Assert.Equal(Encoding.UTF8, sr.CurrentEncoding);
+            }
+        }
```
Comment (stephentoub) [inline src/System.IO/tests/StreamReader/StreamReader.CtorTests.cs]: I'm not understanding what these tests are validating.  How does asserting the current encoding == UTF8 help to validate the behavior of specifying each of these optional arguments?

**Comment by jimdemis** [comment]
The coreclr part is merged. This is ready for review and merge. The `corefx-ci (Packaging All Configurations x64_Debug)` check's error is unrelated. Thanks! /cc @stephentoub @JeremyKuhne

**Comment by stephentoub** [comment]
> This is ready for review and merge.

Thanks, but not until corefx has ingested an updated coreclr.

File: src/System.IO/tests/StreamWriter/StreamWriter.CtorTests.cs [hunk 2]
```diff
@@ -31,15 +31,6 @@ public static void CreateStreamWriter()
...
         }
 
-        [Fact]
-        public static void NullEncodingThrows()
-        {
-            // [] Check for ArgumentNullException on null encoding
-            //-----------------------------------------------------------------
-
-            Assert.Throws<ArgumentNullException>(() => new StreamWriter(new MemoryStream(), null));
-        }
```
Comment (stephentoub) [inline src/System.IO/tests/StreamWriter/StreamWriter.CtorTests.cs]: This should now have the opposite test, validating that null can be passed in. And there should be similar tests for the other optional arguments.

Comment (jimdemis) [reply src/System.IO/tests/StreamWriter/StreamWriter.CtorTests.cs]: Please suggest how I can test this. Should I just use the constructor like this:
```csharp
[Fact]
public void StreamReader_WithOptionalArguments()
{
    using (var sr = new StreamReader(new MemoryStream(), leaveOpen: true)) { }
    using (var sr = new StreamReader(new MemoryStream(), encoding: null)) { }
    using (var sr = new StreamReader(new MemoryStream(), bufferSize: -1)) { }
    using (var sr = new StreamReader(new MemoryStream(), detectEncodingFromByteOrderMarks: true)) { }
}
```

File: src/System.IO/tests/StreamReader/StreamReader.CtorTests.cs (hunk 1 above)
Comment (JeremyKuhne) [reply src/System.IO/tests/StreamReader/StreamReader.CtorTests.cs]: I presume you're trying to make sure the defaults are what you expect when you don't specify specific arguments. This is valuable to do. You can easily check the encoding as you have. It is also relatively easy to check the `leaveOpen` default by looking at `MemoryStream.CanRead` after you dispose the reader. That will be false if it has been disposed.

File: src/System.IO/tests/StreamWriter/StreamWriter.CtorTests.cs (hunk 2 above)
Comment (JeremyKuhne) [reply src/System.IO/tests/StreamWriter/StreamWriter.CtorTests.cs]: You should keep this test and mark it so it only runs on the desktop .NET Framework (e.g. 4.7.2 etc.).  We want to still test the old behavior there. You do this by adding the following attribute:

``` C#
[SkipOnTargetFramework(~TargetFrameworkMonikers.NetFramework)] // e.g. NetFX only
```

File: src/System.IO/tests/StreamReader/StreamReader.CtorTests.cs (hunk 1 above)
Comment (JeremyKuhne) [reply src/System.IO/tests/StreamReader/StreamReader.CtorTests.cs]: As far as `detectEncodingFromByteOrderMarks` goes, it looks like we're weak on testing that in general. I don't see that we have any tests that actually check that DetectEncoding() works as it should. I've created an issue for it: #37190

**Comment by jimdemis** [comment]
Finally, it is green! @JeremyKuhne, @stephentoub, @carlossanlop, @karelz

**Comment by JeremyKuhne** [comment]
> Finally, it is green!

Awesome! Thanks for your patience. :)

**Pull request merged**